src/DGtal/topology/VoxelComplexFunctions.ih

```diff
@@ -0,0 +1,561 @@
+      if (K.findCell(3, it->first) != K.end(3))
+        continue;
+      auto & voxel = it->first ;
+      auto & ccdata = it->second.data ;
+      if( Skel(X, voxel) == true && ccdata == 0)
+        ccdata = generation;
+    }
+    Y = K ;
+    x_y = X; //optimization instead of x_y = X-Y, use x_y -= Y;
+    // d-cliques: From voxels (d=3) to pointels (d=0)
+    for (int d = 3 ; d >= 0 ; --d) {
+      // Search only critical cliques of X in the set X \ Y
+      x_y -= Y; // Y is growing in this d-loop.
+      critical_cliques = X.criticalCliquesForD(d, x_y);
+
+      for(auto & clique : critical_cliques)
+      {
+        // Find the selected voxel in x to get associated Data.
+        auto selected_voxelpair = Select(clique);
+        selected_voxelpair.second = X.findCell(3,selected_voxelpair.first)->second;
+        Y.insertVoxelCell(selected_voxelpair, close_it);
+      }
+    } // d-loop
+    // Y is now  populated with some selected critical cliques for this generation,
+    // plus with the fixed points in K.
+    // Store result (final or for the next generation)
+    X = Y;
+
+    /*** Modify K (the set of voxels to conserve) ***/
+    // Insert in K (inmutable set) only if the NEW critical cells of this generation (Y-K) are:
+    // a) Part of the Skeleton (Skel) user wants to preserve.
+    // b) persistent enough.
+
+    // Update K
+    Y -= K;
+    for (auto it = Y.begin(3), itE = Y.end(3) ; it != itE ; ++it ){
+        auto & voxel = it->first;
+        auto & ccdata = it->second.data;
+        bool is_skel = Skel(X, voxel);
+        bool is_persistent_enough = (generation + 1 - ccdata) >= persistence;
+        if (is_skel && is_persistent_enough)
+          K.insertVoxelCell(voxel, close_it, ccdata);
+    }
+
+    if(verbose){
+      trace.info() << "generation: " << generation <<
+        " ; X  at start generation: " << xsize_old <<
+        " ; K (constraint set): " << K.nbCells(3) <<
+        " ; Y - K: " << Y.nbCells(3) <<
+        " ; Y (X at the end): " << X.nbCells(3) <<
+        std::endl;
+    }
+
+    /***** Stability Update: ****/
+    xsize = X.nbCells(3);
+    if(xsize == xsize_old)
+      stability = true;
+    xsize_old = xsize;
+
+  } while( !stability );
+
+  if(verbose) trace.endBlock();
+
+  return X;
+}
+
+//////////////////////////////////////////////////////////////////////////////
+// Select Functions
+//////////////////////////////////////////////////////////////////////////////
+template < typename TComplex >
+std::pair<typename TComplex::Cell, typename TComplex::Data>
+DGtal::functions::selectFirst(
+      const typename TComplex::Clique & clique)
+{
+  return *clique.begin(3);
+}
+
+template < typename TComplex >
+std::pair<typename TComplex::Cell, typename TComplex::Data>
+DGtal::functions::selectRandom(
+      const typename TComplex::Clique & clique)
+{
+  static std::random_device rd;
+  static std::mt19937 gen(rd()); // TODO tothink provide seed?
```

Comment:
What I do sometimes in this case is to propose providing a whole random generator instead of the seed only, something like that : 
```C++
template < typename TComplex >
std::pair<typename TComplex::Cell, typename TComplex::Data>
DGtal::functions::selectRandom(
      const typename TComplex::Clique & clique)
{
    static std::mt19937 gen( std::random_device{}() );
    return selectRandom(clique, gen);
}

template < typename TComplex, typename TRandomGenerator >
std::pair<typename TComplex::Cell, typename TComplex::Data>
DGtal::functions::selectRandom(
      const typename TComplex::Clique & clique,
      TRandomGenerator & gen
)
{
    auto size = clique.nbCells(3);
    std::uniform_int_distribution<> dis(0, size - 1);
    auto it = clique.begin(3);
    std::advance(it, dis(gen));
    return *it;
}
```

Reply:
Good idea, changing it.